### ISSUE TITLE
When applying a generic `class-string` to `class-string<T>` and `return Foo&T`, psalm infers `mixed`

### ISSUE THREAD

**Issue by Ocramius** [vimeo/psalm#1813]
In the following scenario, since a generic `class-string` is used when calling `mock()`, psalm reduces the declared `@psalm-return RequestedType&MockObject<RequestedType>` (presumably inferred as `@psalm-return object&MockObject<object>`) to just `mixed`.

Ideally, this should be inferred as just `MockObject`, which is the the closest "valid" type:

```php
<?php

class MockObject
{
    public function checkExpectations() : void
    {
    }
}

/**
 * @psalm-template RequestedType
 * @psalm-param class-string<RequestedType> $className
 * @psalm-return RequestedType&MockObject
 * @psalm-suppress MixedInferredReturnType
 * @psalm-suppress MixedReturnStatement
 */
function mock(string $className)
{
    eval('"there be dragons"');

    return $instance;
}

/**
 * @psalm-param class-string $className
 */
function useMock(string $className) : void
{
    mock($className)->checkExpectations();
}
```

https://psalm.dev/r/4785a0a1a5 : 

```
Psalm output (using commit 7155ed4):

INFO: MixedMethodCall - 29:23 - Cannot determine the type of the object on the left hand side of this expression
```

EDIT: simplified example.

**Comment by weirdan** [comment]
Note that it's order-dependent. If you spell return type as `MockObject<RequestedType>&RequestedType` Psalm does not emit any errors: https://psalm.dev/r/9d850ddda0

**Comment by Ocramius** [comment]
Hmm, this has bitten me too many times then: why is there a concept of order dependency in thr type there?

**Comment by muglug** [comment]
There are no concepts of order-dependence, only bugs ;)
